src/llvm/builders/To.js

```diff
@@ -0,0 +1,187 @@
+  IRBuilder,
+  UndefValue,
+} from "llvm-node";
+import { llvmContext } from "../llvm-context.js";
+
+import { buildFromStringValue, getStringLength } from "./StringValue.js";
+
+export function buildToBoolean(state: CompilerState, value: LLVMValue, builder: IRBuilder): LLVMValue {
+  if (value.type.isIntegerTy(1)) {
+    // Boolean
+    // Already boolean
+    return value;
+  } else if (value.type.isDoubleTy()) {
+    // Number
+    // NaN is false
+    return builder.createFCmpONE(value, ConstantFP.get(llvmContext, 0));
+  } else if (value.type.isIntegerTy(32)) {
+    // Signed Integral
+    return builder.createICmpNE(value, ConstantInt.get(llvmContext, 0));
+  } else if (state.intrinsics.isUint32Type(value.type)) {
+    // Unsigned Integral
+    let unsignedValue = builder.createExtractValue(value, [0]);
+    return builder.createICmpNE(unsignedValue, ConstantInt.get(llvmContext, 0));
+  } else if (state.intrinsics.isStringType(value.type)) {
+    // String
+    // True of non-zero length.
+    return builder.createICmpNE(getStringLength(value, builder), ConstantInt.get(llvmContext, 0));
+  } else {
+    invariant(false, "only string, boolean and numbers should need toBoolean");
+  }
+}
+
+export function buildToNumber(state: CompilerState, value: LLVMValue, builder: IRBuilder): LLVMValue {
+  if (value.type.isIntegerTy(1)) {
+    // Boolean
+    return builder.createSelect(value, ConstantFP.get(llvmContext, 1), ConstantFP.get(llvmContext, 0));
+  } else if (value.type.isDoubleTy()) {
+    // Number
+    // Already a number
+    return value;
+  } else if (value.type.isIntegerTy(32)) {
+    // Signed Integral
+    return builder.createSIToFP(value, LLVMType.getDoubleTy(llvmContext));
+  } else if (state.intrinsics.isUint32Type(value.type)) {
+    // Unsigned Integral
+    let unsignedValue = builder.createExtractValue(value, [0]);
+    return builder.createUIToFP(unsignedValue, LLVMType.getDoubleTy(llvmContext));
+  } else if (state.intrinsics.isStringType(value.type)) {
+    // String
+    // TODO: Call library function to parse a string.
+    let error = new CompilerDiagnostic(
+      `ToNumber from String is not yet implemented for LLVM.`,
+      undefined,
+      "PP2000",
+      "FatalError"
+    );
+    state.realm.handleError(error);
+    throw new FatalError();
+  } else if (value.type.isVoidTy() || (value instanceof Constant && value.isNullValue())) {
+    // Undefined or Null
+    return ConstantFP.get(llvmContext, 0);
+  } else {
+    invariant(false, "only string, boolean and numbers should need toNumber");
+  }
+}
+
+export function buildToInt32(state: CompilerState, value: LLVMValue, builder: IRBuilder): LLVMValue {
+  if (value.type.isIntegerTy(1)) {
+    // Boolean
+    return builder.createSelect(value, ConstantInt.get(llvmContext, 1), ConstantInt.get(llvmContext, 0));
+  } else if (value.type.isDoubleTy()) {
+    // Number
+    // TODO: I think we can make this fewer instructions/faster by some clever bit manipulation.
```

Comment:
Why the special handling of infinity/NaN? Reading the spec, FPToSI would result in something undefined for all out-of-range values, not just infinity/NaN.

Reply:
This is the internal operation called by `x | 0` yes. `<<`, `>>` and `>>>` uses ToUint32 which is slightly different.

https://tc39.github.io/ecma262/#sec-toint32

I'm not sure what you're asking. The spec requires infinity/nan to have a certain behavior. The FPToSI behavior is somewhat undefined but will in practice simply yield the wrong result so I can't rely on it alone.

The mod 2^32 is speced because it basically means taking the first 32 bits of the fraction after the rounding. This is different than the overflow behavior you would get from using any of the other LLVM operations.

The reason things are defined this way is so that multiple operations are idempotent which helps when you combine multiple of these operations in sequence and asm.js. I don't fully understand it but seems clever.

None of this applies to any single or even few CPU instructions. Interestingly no implementation does the same thing here. Everyone tries their own clever thing. I can't rely on certain platform specific quirks so not all solution would work cross-platform. I have some ideas around 64-bit casting (not currently exposed by the llvm-node bindings). In the end I opted for just the simplest code.